[Sukkaito] iP

Sukkaito opened the pull request.

Sukkaito committed:
Rename to SuBOT, add basic greets and exits

Sukkaito committed:
Add Level-1 incremement

Sukkaito committed:
Add command history, Level-2 increment

Sukkaito committed:
Add Level-3 increment

Changes:
- "list" command no longer shows command history, but shows list of tasks instead.

Feature:
- Ability to mark/unmark task as _done_
- Optional ```Task``` class

Sukkaito committed:
Follow coding convention

rharwo commented on src/main/java/SuBOT.java:
```diff
@@ -0,0 +1,73 @@
+import java.util.Scanner;
+
+public class SuBOT {
+    private static boolean stopFlag = false;
```
```suggestion
    private static boolean stopFlag = isRunning;
```

Sukkaito replied:
Thank you for the suggestion! Added! 👍

rharwo commented on src/main/java/SuBOT.java:
```diff
@@ -0,0 +1,73 @@
+
+public class SuBOT {
+    private static boolean stopFlag = false;
+    private static Task[] tasks = new Task[100];
```
Is it wise to use magic numbers? Perhaps you could make it a final variable.

Sukkaito replied:
Thank you for the suggestion! Added! 👍

rharwo commented on src/main/java/SuBOT.java:
```diff
@@ -0,0 +1,73 @@
+     * @param command Input command
+     */
+    public static void process(String command) {
+        String[] argv = command.split(" ");
+        int argc = argv.length;
```
Perhaps you could use more descriptive names than argv and argc.

Sukkaito replied:
I think `argc` and `argv` are already standard names here (at least in C++). It shouldn't be a problem.

rharwo commented on src/main/java/SuBOT.java:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    public static void main(String[] args) {
+        String logo = """
+                       ____  ____  ______
+           _______  __/ __ )/ __ \\/_  __/
+          / ___/ / / / __  / / / / / /
+         (__  ) /_/ / /_/ / /_/ / / /
+        /____/\\__,_/_____/\\____/ /_/
+        """;
```
Perhaps change the logo to a final variable if it is not going to change.

Sukkaito replied:
Thank you for the suggestion! Added! 👍

rharwo commented on src/main/java/SuBOT.java:
```diff
@@ -0,0 +1,73 @@
+
+        System.out.println(logo);
+        System.out.println("How can I help you today?");
+        System.out.println("-------------------------------");
```
Could you use a final variable so that you do not have to keep repeating "----" in your code?

Sukkaito replied:
Thank you for the suggestion! Added! 👍

rharwo reviewed (commented):
Looks good overall, some minor code quality fixes.

Sukkaito committed:
Add 3 new types of tasks: ToDos, Events, Deadlines

Features:
- ToDos: tasks without any date/time attached to it
- Deadlines: tasks that need to be done before a specific date/time
- Events: tasks that start at a specific date/time and ends at a specific date/time
*Note: date/time is treated as `String`

Changes:
- `Task` class now can only create through chid class
- Only accept specific commands, others will be ignored

Sukkaito committed:
Add automated testing

Features:
- Automated testing

Sukkaito committed:
Improve code quality. Thank you @rharwo for PR reviews

Changes:
- Extract `process()` to a new class, `ActionManager` and extract further from there.
- Follow coding standards and conventions.
- Javadoc support.

Sukkaito committed:
Add custom exceptions and handlers

Features:
- Custom exceptions
- Custom responses on invalid command/arguments

Changes:
- Improve code quality with `Parser` class

Sukkaito committed:
Add custom test cases

Features:
- Test cases for every command

Sukkaito committed:
Merge branch 'branch-Level-5'

Sukkaito committed:
Fix: Correct expected output for input2.txt (`event` command test)

Kobot7 commented on src/main/java/ActionManager.java:
```diff
@@ -0,0 +1,163 @@
+            tasks[taskIndex].setDone(isMark);
+            System.out.printf("I have %sed the following task for you:\n", action);
+            tasks[taskIndex].printTask();
+        } catch (SuBotException e) {
```
Is the variable name "e" informative enough?

Kobot7 commented on src/main/java/ActionManager.java:
```diff
@@ -0,0 +1,163 @@
+    }
+
+    /**
+     * Marks/unmarks a task as done
```
Leave an empty line between description and parameter section

Kobot7 commented on src/main/java/ActionManager.java:
```diff
@@ -0,0 +1,163 @@
+
+    /**
+     * Marks/unmarks a task as done
+     * @param action <code>mark</code> or <code>unmark</code>
```
Ensure that there is punctuation behind each parameter description

Kobot7 commented on src/main/java/ActionManager.java:
```diff
@@ -0,0 +1,163 @@
+            case "list":
+                listAllTask();
+                break;
+            case "mark":
```
Is there meant to be a "break" here? Otherwise, it would be good to indicate that this was intended using "//Fallthrough"

Kobot7 commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,46 @@
+        this.time = time;
+    }
+
+    @Override
+    public String getCommand() {
+        return super.getCommand();
+    }
+
+    @Override
+    public void setCommand(String command) {
+        super.setCommand(command);
+    }
+
+    @Override
+    public boolean isDone() {
+        return super.isDone();
+    }
+
+    @Override
+    public void setDone(boolean done) {
+        super.setDone(done);
+    }
```
Inherited methods do not need to be redeclared if there are no changes to their functionality

Kobot7 commented on src/main/java/Parser.java:
```diff
@@ -0,0 +1,92 @@
+            throw new SuBotException("Task number must be a positive integer");
+        }
+    }
+
+    /**
+     * Parse the input arguments for {@link ActionManager#addEvent(String) addEvent} method
+     * @param args Arguments received from upstream stored in a <code>String</code>
+     * @return Array of arguments: <code>{eventName, eventFrom, eventTo}</code>
+     * @throws SuBotException If any missing arguments exists or invalid format
+     */
+    static String[] parseEvent(String args) throws SuBotException {
+        int fromSlashIndex = args.indexOf("/from");
+        if (fromSlashIndex == -1)  {
+            throw new SuBotException("\"/from\" expected");
+        }
+
+        int toSlashIndex = args.indexOf("/to");
+        if (toSlashIndex == -1)  {
+            throw new SuBotException("\"/to\" expected");
+        }
+
+        String eventName = args.substring(0, fromSlashIndex).strip();
```
Consider using a named constant instead of a magic number

Kobot7 commented on src/main/java/Parser.java:
```diff
@@ -0,0 +1,92 @@
+            int taskIndex = Integer.parseUnsignedInt(args.trim()) - 1; //Index starts from 0
+            if (taskIndex < 0) {
+                throw new SuBotException("Task number must be a positive integer");
+            } else { return taskIndex; }
```
Follow the layout convention for if-else statements, format it using Egyptian style

Sukkaito committed:
Put classes into separated packages

Sukkaito committed:
Merge branch 'branch-A-Packages'

Sukkaito committed:
Fix: Typo in ActionManager.js

Sukkaito committed:
Add ability to delete task from the list

Features:
- `delete` command: `delete <taskNumber>`

Changes:
- Rename `parseMarkUnmark()` to `parseTaskIndex()` for reusability
- Task list now stored in `ArrayList<Task>` instead of `Task[]`

Sukkaito committed:
Add tests for `delete` command

Sukkaito committed:
Feature: Save and load tasks from disks

Feature:
- Autoload tasks from previous session from file on startup
- Autosave tasks from current session to disk

Sukkaito committed:
Ignore local save file

Sukkaito committed:
Merge branch 'branch-Level-6'

Sukkaito committed:
Merge branch 'branch-Level-7'

Sukkaito committed:
Update tests for new implements

Sukkaito committed:
Refactor the code to different classes

Sukkaito committed:
Feature: `find` command

Feature:
- Add command `find`: Find tasks based on the given query

Sukkaito committed:
Merge pull request #1 from Sukkaito/branch-Level-9

Feature: `find` command

Sukkaito committed:
Update JavaDoc

Changes:
- Add JavaDoc support for new classes

Sukkaito committed:
Merge pull request #2 from Sukkaito/branch-A-JavaDoc

Update JavaDoc

Sukkaito committed:
Update user guide
